pid1: stop refusing to boot with cgroup v1

keszybz opened the pull request:
Since v256 we completely fail to boot if v1 is configured. Fedora 41 was just released with v256.7 and this is probably the first major exposure of users to this code. It turns out not work very well. Fedora switched to v2 as default in F31 (2019) and at that time some people added configuration to use v1 either because of Docker or for other reasons. But it's been long enough ago that people don't remember this and are now very unhappy when the system refuses to boot after an upgrade.

Refusing to boot is also unnecessarilly punishing to users. For machines that are used remotely, this could mean somebody needs to physically access the machine. For other users, the machine might be the only way to access the net and help, and people might not know how to set kernel parameters without some docs. And because this is in systemd, after an upgrade all boot choices are affected, and it's not possible to e.g. select an older kernel for boot. And crashing the machine doesn't really serve our goal either: we were giving a hint how to continue using v1 and nothing else.

So let's boot after a delay, and actually give a hint how to switch to v2. And if people do the override for v1, boot without delay. We made the point, no need to punish people further.

https://bugzilla.redhat.com/show_bug.cgi?id=2323323 https://bugzilla.redhat.com/show_bug.cgi?id=2323345 https://bugzilla.redhat.com/show_bug.cgi?id=2322467 https://www.reddit.com/r/Fedora/comments/1gfcyw9/refusing_to_run_under_cgroup_01_sy_specified_on/

The advice is to set systemd.unified_cgroup_hierarchy=1 (instead of removing systemd.unified_cgroup_hierarchy=0). I think this is easier to convey. Users who are understand what is going on can just remove the option instead.

YHNdnzj commented:
I'm not convinced really, especially considering the fact that it's very effective in the Arch and probably NixOS community from my observations. It also encouraged e.g. Kubernetes to move towards v2. Furthermore, we'd rip it out completely in v258, hence taking a step back at this point does not help either.

My suggestion would be to detect this at package upgrade time and probably even automatically enable v2 and emit a big fat warning on this. Yes it's intrusive, yet the removal is impending and failing to act in time would only cause more user confusion in the next release.

YHNdnzj commented:
IOW, I see this as a chance for distros to figure out viable migration paths for normal users. If they are never aware of the deprecation even after 2 years, what happens when v1 is removed completely? 

It just came to my notice that Gentoo has been doing what I'm proposing: https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=2e0a4dbc6b7753f19a8a6243603bd2792082359e

keszybz commented:
Arch and Nixos are niche distributions and the users of those distributions are accustomed to the requirement  of frequent manual interventions. But we also need to care about more mainstream distributions where the expectation is that upgrades "just work". Breaking boot is just a wrong thing to do almost always. We could consider doing that in some extreme cases, if we have no way of reasonably proceeding. But the fact that we plan to remove some functionality in a future release is not a good reason.

> I see this as a chance for distros to figure out viable migration paths for normal users.

Exactly. I think the only reasonable answer is that we'll switch to v2 automatically and simply ignore (possibly with a warning) all configuration that says that v1 should be used. If we accept that that will be end state, then I think it's reasonable to move smoothly towards that state by issuing a warning now.

>If they are never aware of the deprecation even after 2 years, what happens when v1 is removed completely?

I think you're confusing distributions and users. Distributions are aware and are moving towards v2. But individual users are much more varied. Most users don't really care and want things to continue to work with the minimum of fuss. This is what this patch does.

YHNdnzj commented:
> I think the only reasonable answer is that we'll switch to v2 automatically and simply ignore (possibly with a warning) all configuration that says that v1 should be used.

I strongly disagree. First of all, it's inherently wrong to silently discard a system-wide configuration. We need to inform the user in some ways, even if we decide to automatically flip it for them. Minimizing confusion matters more than tentative disturbance in my eyes.

And secondly - how are we supposed to handle daemon-reexec to v258 on cgroup v1 systems? This _requires_ proactive intervention from distros, and Gentoo sets a great example here.

keszybz commented:
> > I think the only reasonable answer is that we'll switch to v2 automatically and simply ignore (possibly with a warning) all configuration that says that v1 should be used.
> 
> I strongly disagree. First of all, it's inherently wrong to silently discard a system-wide configuration. We need to inform the user in some ways, even if we decide to automatically flip it for them. Minimizing confusion matters more than tentative disturbance in my eyes.

Hmm, you say you disagree, but then you say that we should do pretty much what I wrote.

> And secondly - how are we supposed to handle daemon-reexec to v258 on cgroup v1 systems? This _requires_ proactive intervention from distros, and Gentoo sets a great example here.

Gentoo's approach is suitable for gentoo, where users build their own packages. On binary distributions the build a different approach must be used. The details how this is best done will vary between distributions. For example, Debian-family distros usually do interactive package updates and it might be suitable to get user input during the upgrade. Rpm distros do noninteractive package upgrades, so they want the upgrade to do something sensible automatically. The system upstream should allow different distributions to do what is appropriate for them, which means that we must not actually break boot. Different distros will take care to inform the user in the most appropriate fashion that matches their style.

YHNdnzj commented:
> > > I think the only reasonable answer is that we'll switch to v2 automatically and simply ignore (possibly with a warning) all configuration that says that v1 should be used.
> > 
> > I strongly disagree. First of all, it's inherently wrong to silently discard a system-wide configuration. We need to inform the user in some ways, even if we decide to automatically flip it for them. Minimizing confusion matters more than tentative disturbance in my eyes.
> 
> Hmm, you say you disagree, but then you say that we should do pretty much what I wrote.

Ugh right, I was trying to say that the "flip" should be done by distro in a way that's obvious to the user, not by upstream. I.e. in v258 I'd still prohibit kernel cmdline option that enables cg v1.

> The system upstream should allow different distributions to do what is appropriate for them, which means that we must not actually break boot. Different distros will take care to inform the user in the most appropriate fashion that matches their style.

Still, it will naturally and forcibly break boot in v258. Anyways, all I'm saying is that I tend to treat the prevention of boot as a "catch-all" method. It serves the purpose by covering the case for e.g. Arch where manual intervention is anticipated, but also revealed Fedora's problem (allow me to put it that way). Fedora doesn't need to disable the mechanism to accomplish the smooth upgradation IMHO. Quite the opposite, the conversion must be at the distro/pkg manager level, before systemd is involved (daemon-reexec). Frankly, if the migration is done properly now, upstream disabling cg v1 shouldn't become an issue in the first place, no?

poettering commented:
What about just booting into cgroupv2 mode now, even if cgroupv1 is requested, unless SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1 is set?

So you'd have three options:

1. Boot quickly into cgroupv2 mode
2. Boot into cgroupv1 mode, but wait for an annoying 30s and then actually end up in cgroupv2 mode anyway.
3. Boot into cgroupv1 mode, because you set SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1 and like to live in the past.

I think that would be an OK way out?

keszybz commented:
Updated as suggested by @bluca  and @poettering. IIUC, both describe the same approach.

poettering commented:
Oh, I am seeing only now that @bluca actually suggested the same thing as I did...

poettering commented:
Hmm, I sense this should probably be documented somewhere, i.e. NEWS (it mentions SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE a couple of times)

keszybz commented:
Sorry, this needs a rework. I'll update shortly.

bluca commented on src/shared/mount-setup.c:
```diff
@@ -512,11 +512,21 @@ int mount_cgroup_legacy_controllers(bool loaded_policy) {
-        if (!cg_is_legacy_force_enabled())
-                return -ERFKILL;
+        if (cg_is_legacy_force_enabled())
+                log_warning("Legacy support for cgroup v1 enabled via SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1.");
```
I think this should also include the same "this will stop being supported" wording, to try and communicate that this workaround is not permanent

keszybz replied:
It says "temporarily". I amended NEWS now as requested by Lennart.

keszybz replied:
Oh, it says so in the other branch. But I think here we don't need this. We are very clear that it's temporary in other places, and once the user has made their decision to add the override, we don't need to repeat ourselves.

YHNdnzj commented on src/shared/cgroup-setup.c:
```diff
@@ -152,19 +158,23 @@ bool cg_is_hybrid_wanted(void) {
         /* If in container, we have to follow host's cgroup hierarchy. */
         if (detect_container() > 0)
-                return true;
+                return (enabled = true);
```
This seems wrong. Previously `cg_is_legacy_wanted` is checked only after `cg_is_unified_wanted`, i.e. the former returns true as generic fallback. Now you're effectively enabling v1 for all containers since _unified_wanted calls _legacy_wanted first...

keszybz commented:
OK, updated to take into account @YHNdnzj's comment.

Tested with:
- a VM boot without any options
- a VM boot with `systemd.unified-cgroup-hierarchy=0` → 30 s delay, cgroup v2
- a VM boot with `systemd.unified-cgroup-hierarchy=0 SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1` → no delay, v1
- systemd-nspawn on a container, host has v2 → v2

So I think this is good to go, but since the code changed a bit again, PTAL.

behrmann commented:
The delay of 30s is quite friendly I must say. For baremetal servers this is small against time spent in firmware and when not actively looking at a machine boot one might easily miss it while getting a coffee.

keszybz commented:
Jesus, this still doesn't behave correctly in all cases. I'll need to do an update.

keszybz commented:
Updated again, PTAL.

I dropped caching in two places: in `cg_is_legacy_wanted()` and `cg_is_legacy_enabled_force()` (only added in previous version of the patch). I also restored `log_full(LOG_CRIT)` for the warning. It doesn't matter much for how the message is displayed, but maybe it'll pass some more filters, and we want it to be as visible as possible. I also changed/simplified the logic a bit.



> The delay of 30s is quite friendly I must say. For baremetal servers this is small against time spent in firmware and when not actively looking at a machine boot one might easily miss it while getting a coffee.

It's actually a minute, because we get the warning once in the initrd and once on the host system once systemd has been updated in both places. We *could* get rid of the duplication, but having it in both places makes it more likely to be seen, and some people use don't use systemd in the initrd, etc. We could add code to propagate information that we've already displayed the warning, but I don't think we want the complexity, and maybe it's better to have it twice anyway.

The warning might be missed during boot, but note that we change to v2 when the warning is displayed. Either people are fine with that and they don't care, or they actually depend on v1 being used. If things break for them, they'll look at the logs and see the issue. I don't think that there's any particular delay that'd make a difference. During interactive boot, 30 s is likely to be noticed. But if somebody is not watching the machine, even 10 minutes can be easily missed. So I think the current values are reasonable.

keszybz commented:
I forgot to add: this setup code is a fragile mess. I'll be very happy when we get rid of it.

poettering commented on src/shared/cgroup-setup.c:
```diff
@@ -92,11 +92,16 @@ bool cg_is_unified_wanted(void) {
-        /* If we were explicitly passed systemd.unified_cgroup_hierarchy, respect that. */
+        /* If we have explicit configuration for v1 or v2, respect that. */
+        if (cg_is_legacy_force_enabled()) {
+                log_notice("cg_is_legacy_force_enabled -> true");
```
leftover debug message

keszybz replied:
Arrgh. Updated.

keszybz commented:
I repeated the same tests:

- a VM boot without any options → v2, no delay or warnings
- a VM boot with `systemd.unified-cgroup-hierarchy=0` → 30 s delay, warning, cgroup v2
- a VM boot with `systemd.unified-cgroup-hierarchy=0 SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1` → no delay, v1
- systemd-nspawn on a container, host has v2 → v2

bluca commented:
lxc is started on ubuntu 20.04, so I guess that's booted with the legacy hierachy, and I guess this introduces a regression when ran as container payload?

keszybz commented:
Updated once again. I restored the check for containers. The patch looks very ugly because we first do the the cgroup v2 mounts and get the cgroup v1 check later. But I don't think it'd be useful to completely rework this code since we plan to remove it.

Tested with:
- a VM boot without any options → v2, no delay or warnings
- a VM boot with systemd.unified-cgroup-hierarchy=0 → 2×30 s delay and warning, cgroup v2
- a VM boot with systemd.unified-cgroup-hierarchy=0 SYSTEMD_CGROUP_ENABLE_LEGACY_FORCE=1 → no delay, v1, warning
- systemd-nspawn on a container, host has v2 → v2
- systemd-nspawn on a container, host has v1 → v1, warning

I didn't text lxc, I'll rely on the CI to do that.

keszybz commented:
noble/amd64 is not my fault:
 92s OSError: [Errno 28] No space left on device
 92s autopkgtest [12:41:19]: ERROR: testbed failure: unexpected eof from the testbed
